lib/OpenQA/WebAPI/Plugin/ObsRsync/Task.pm

```diff
@@ -0,0 +1,154 @@
+use constant {
+    QUEUED     => 200,
+    STARTED    => 201,
+    IN_QUEUE   => 208,
+    QUEUE_FULL => 507,
+};
+
+sub register {
+    my ($app) = @_;
+    my $home = $app->config->{obs_rsync}->{home};
+    return undef unless $home;
+
+    my $concurrency        = $app->config->{obs_rsync}->{concurrency} // 2;
+    my $project_status_url = $app->config->{obs_rsync}->{project_status_url};
+    my $retry_interval     = $app->config->{obs_rsync}->{retry_interval} // 15;
+    my $lock_timeout       = 36000;
+
+    my $minion = $app->minion;
+
+    $minion->add_task(
+        obs_rsync_queue => sub {
+            my ($job, $args) = @_;
+            my $project = $args->{project};
+
+            my $results
+              = $minion->backend->list_jobs(0, undef, {tasks => ['obs_rsync_run'], states => ['active', 'inactive']});
+            for my $other_job (@{$results->{jobs}}) {
+                if ($other_job->{args} && ($other_job->{args}[0]->{project} eq $project)) {
+                    return $job->finish(IN_QUEUE) if $other_job->{notes}{waitingconcurrencyslot};
+                    return $job->retry({delay => $retry_interval});
+                }
+            }
+            $app->gru->enqueue(
+                'obs_rsync_run',
+                {project  => $project},
+                {priority => 100, notes => {waitingconcurrencyslot => 1}});
+            return $job->finish(STARTED);
+        });
+
+    $minion->add_task(
+        # actual rsync gru task
+        obs_rsync_run => sub {
+            my ($job, $args) = @_;
+            my $project = $args->{project};
+
+            return $job->retry({delay => $retry_interval}) if _is_obs_status_dirty($project_status_url, $project);
+            return $job->retry({delay => $retry_interval})
+              unless my $guard = $minion->guard('obs_rsync_run_guard', $lock_timeout, {limit => $concurrency});
+
+            $job->note(waitingconcurrencyslot => undef);
+            eval { system([0], "bash", "$home/rsync.sh", $project); };
+            return $job->finish($EXITVAL);
+        });
+}
+
+sub schedule {
+    my ($controller, $project) = @_;
+    my $app = $controller->app;
+
+    my $results = $app->minion->backend->list_jobs(0, undef,
+        {tasks => ['obs_rsync_run', 'obs_rsync_queue'], states => ['active', 'inactive']});
+
+    for my $other_job (@{$results->{jobs}}) {
+        if ($other_job->{args} && ($other_job->{args}[0]->{project} eq $project)) {
+
+            return $controller->render(status => IN_QUEUE)
+              if ($other_job->{task} eq 'obs_rsync_queue' || $other_job->{notes}{waitingconcurrencyslot});
+            $app->gru->enqueue('obs_rsync_queue', {project => $project}, {priority => 90});
+            return $controller->render(json => {message => "ok"}, status => QUEUED);
+        }
+    }
+    my $queue_limit = $app->config->{obs_rsync}->{queue_limit} // 100;
+    return $controller->render(status => QUEUE_FULL) if ($results->{total} >= $queue_limit);
```

Comment:
They indeed could be simpler if rsync was always performed immediately after request from cient. Since we have to honour concurrency limit (to avoid overload of IO system) - we must properly stack subsequent requests for the same project and honour project status in OBS.

Reply:
I don't understand this part. Concurrency limits should be much easier to enforce with Minion locks. The `list_jobs` stuff should not be necessary.

Reply:
But stacking too many requests doesn't make sense, thus we need some limit when we honestly say that we give up.

Concurrency indeed is controlled by Minion guard. But for subsequent rsync request on the same project we must know whether existing job really started rsync or jas job which is still waiting for concurrency slot. That info is stored in job->notes and is used to determine whether we queue up new job or do nothing, considering that existing queue will satisfy this request.